### ISSUE TITLE
Antex called twice, once without node-specific config

### ISSUE THREAD

**Issue by jonsterling** [paolobrasolin/jekyll-sheafy#13]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

I'm noticing some behavior which might be a bug in jekyll-sheafy or in jekyll-antex or in both. If you have a node with some embedded TeX that is meant to be passed through Antex, this causes **two** build tasks for Antex:

1. Build the TeX code **with** the node's local `antex: ...` configuration.
2. Build the TeX code **without** the node's local `antex: ...` configuration (so, only using the global one).

Previously I didn't notice this issue because I was only using a global jekyll-antex configuration.

### Expected behaviour

My main priority for this ticket is: jekyll-antex should always be activated with the current node's local configuration; secondarily, it seems also that it should only be run once.

### Steps to reproduce

1. Set a local configuration in your own node `0001.md` like:
````yaml
antex:
  preamble: >-
    \usepackage{some-additional-package}
````

2. Build it, and look at the `.antex-cache` folder: there will be two `.tex` files that have the same body but different preambles.

### Additional context

Fixing this problem would enable me to make a single unified website for a number of different translation projects, etc. Right now I am prevented from doing that, because I need to have different macros in different "lectures".

**Comment by jonsterling** [comment]
I have investigated this matter a bit further. What appears to be happening is: the SVG is regenerated for *every* time a given node is `@include`'d, and this regeneration always uses the `antex` settings that are attached to the outer page being generated. So one way to test this is to create a node tree like `root > 0001 > 0002` and have an antex diagram in `0002`. Then set different antex settings on `root` and `0001` and `0002` (e.g. with a macro that is defined differently in each one). The result will be three different diagrams.

I think that the desired behavior is as follows:

1. The svg should be generated only once, at the *leaf* node. This svg should be used whenever that node is included.
2. Antex settings inheritance should percolate downward from the root to the leaves. (I believe this is sensible because of the invariants on the node tree.)

**Comment by jonsterling** [comment]
Incidentally, my own user-land implementation of KaTeX rendering and macro loading was incorrect in the identical way (the behavior ignored the macro library settings of the inner node when it is rendered inside another node). I have fixed that here: https://github.com/jonsterling/math/commit/a3455d17e5f545f98ac00777136dc8b862f09e88

**Comment by paolobrasolin** [comment]
Spot-on analysis! Thanks @jonsterling.

This is a _fun_ bug. It's an unexpected interaction between `jekyll-sheafy` and `jekyll-antex`:

1. `jekyll-sheafy` flattens and includes subtrees in a straightforward way that simply ignores the frontmatters of the children nodes: https://github.com/paolobrasolin/jekyll-sheafy/blob/6c4062b522b86c7ad32b8eaa6362e38f4177a629/lib/jekyll/sheafy/dependencies.rb#L148-L155
2. `jekyll-antex` is run _after_ that step, so each snippet ends up compiling multiple times. In fact, compilation jobs are indexed by the MD5 hash of their code, and different MD frontmatters give different Tex preambles.

> Antex settings inheritance should percolate downward from the root to the leaves

Some thoughts:
1. whatever we do, we should avoid coupling `jekyll-sheafy` and `jekyll-antex`
2. downwards percolation could make sense, if implemented as a generic global configuration for `jekyll-sheafy` (i.e. "dear sheafy please percolate `antex.preamble`, `antex.something` and `foobar.qux`")
3. we could also go the other way around and preprocess snippets in leaves inlining `antex` configuration from their preamble (`jekyll-antex` allows that) and _only then_ flatten the tree.
4. 3 sounds nice in theory but It could get messy quick as it still allows parent nodes to override _something_ in the snippets of children nodes. Maybe this suggests that the we should simply NOT allow `antex` configuration anywhere except in the root note? (again, this could be a global `jekyll-sheafy` config).

To sum up, I can imagine three alternatives:
2. allow restricted percolation config and preprocess the frontmatters to push some keys down the tree;
3. inline preambles into snippets during preprocessing and just let `jekyll-antex` do its thing;
4. allow forbidden non-root keys config to force the antex configuration to be global (either at site level or at root level).

I'm biased towards 4 because (it's easier to implement and) it "makes sense" to have a global preamble when writing a book.
I'll sleep a bit on this, I'm not quite sure what's our best option.

**Comment by jonsterling** [comment]
Thanks for looking into this! I just wanted to update you that in the meanwhile, I did end up not trying to configure antex locally at all --- the basic change is that I manually specify the macro library an argument each time I make a `{%tex%}` block. I don't love this (I feel it breaks an abstraction), but I understand the trade-offs that might make the other options have much worse trade-offs in terms of implementation modularity, etc.
